designcompose/src/main/java/com/android/designcompose/squoosh/SquooshRender.kt

```diff
@@ -0,0 +1,288 @@
+
+                                // We're starting a mask operation, so save a layer, and go on to
+                                // render children. If we encounter another mask, or if we get to
+                                // the end of the children, then we need to pop the mask.
+                                pendingMask = child
+                                child = child.nextSibling
+
+                                drawContext.canvas.saveLayer(nodeSize.toRect(), Paint())
+                            } else {
+                                val childLayout = child.computedLayout
+                                if (childLayout != null) {
+                                    translate(
+                                        childLayout.left * density,
+                                        childLayout.top * density
+                                    ) {
+                                        renderNode(child!!)
+                                    }
+                                }
+                                child = child.nextSibling
+                            }
+                        }
+
+                        // XXX: This logic is duplicated above; it needs to be factored out
+                        //      somehow.
+                        if (pendingMask != null) {
+                            val dstInPaint = Paint()
+                            dstInPaint.blendMode = BlendMode.DstIn
+
+                            // Draw the mask as DstIn
+                            drawContext.canvas.saveLayer(nodeSize.toRect(), dstInPaint)
+                            translate(
+                                pendingMask.computedLayout!!.left * density,
+                                pendingMask.computedLayout!!.top * density
+                            ) {
+                                renderNode(pendingMask)
+                            }
+
+                            drawContext.canvas.restore()
+
+                            // Restore the layer that got saved for the mask and content.
+                            drawContext.canvas.restore()
+                        }
+                    }
+                    nodeRenderCount++
+                }
+                renderNode(node)
+            }
+            if (renderTime > 16)
+                Log.d(TAG, "$docName rendered $nodeRenderCount nodes in ${renderTime}ms")
+        }
+    )
+
+@OptIn(ExperimentalTextApi::class)
+private fun squooshTextRender(
+    drawContext: DrawContext,
+    density: Density,
+    textInfo: TextMeasureData,
+    style: ViewStyle,
+    computedLayout: Layout,
+    customizations: CustomizationContext,
+    nodeName: String,
+) {
+    val paragraph =
+        Paragraph(
+            paragraphIntrinsics = textInfo.paragraph,
+            width = computedLayout.width * density.density,
+            maxLines = textInfo.maxLines,
+            ellipsis = style.text_overflow is TextOverflow.Ellipsis
```

Comment:
I've got a local fix for this (it looks at whether a line exceeds the layout allocated bounds after layout is complete and computes a new `maxLines`) but I implemented it in the wrong place (in rendering, rather than in a post-layout pass) and I'd rather just land Squoosh as-is now and then fix bugs.